edtlib: add lookup table for nodes from their dependency ordinals

mbolivar-nordic opened the pull request:
Reworked version of https://github.com/zephyrproject-rtos/zephyr/commit/f2c0e648c0baa4679e4335514114dac26c699881.

Taken from / needed by #26616.

mbolivar-nordic committed:
edtlib: propertize EDT.scc_order, set up graph earlier

Make the scc_order method a property instead. This is in keeping with
the "General biased advice" at the top of file.

The actual order is therefore lazily initialized in this commit and
the order is not computed by the time __init__() returns. The next
commit will invoke scc_order by the time the constructor returns.

This is preparation work for adding a lookup table from dependency
ordinals to nodes. The combination of these two changes will make
intializing that lookup table a bit easier.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

pabigot committed:
edtlib: add EDT.dep_ord2node

Add a lookup table for finding a node by its dependency ordinal.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

pabigot reviewed (approved):
Typo in the commit message of edtlib: add EDT.depord2node: s/looup/lookup/

Otherwise looks great; thanks.

mbolivar-nordic commented:
> Typo in the commit message of edtlib

Fixed, thanks.

galak commented on scripts/dts/edtlib.py:
```diff
@@ -134,6 +134,10 @@ class EDT:
       A collections.OrderedDict that maps a node label to the node with
       that label.
 
+    depord2node:
```
Can we come up with a better name?  Or dep_ord2node?

pabigot replied:
I don't care what it's named.

mbolivar-nordic replied:
Rebased and renamed `dep_ord2node`.

galak reviewed (commented):
LGTM, one query if we can find a better name for depord2node.

mbolivar-nordic committed:
dts: break cycles in test scripts

We don't want to support cyclic dependency structures, because it
means that Node objects cannot have dep_ordinal attributes as they are
currently documented to possess unconditionally.

Nevertheless, we have some in our tests. Remove them by extracting the
/props/ctrl-X nodes to the same level as the /props nodes. This breaks
a cycle caused by:

- /props/ctrl-X nodes depend on /props because of the parent/child
  relationship
- /props depends on /props/ctrl-X because it refers to them by phandle

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
The pytest-ified test suite seems to have exposed cycles in our test case devicetrees that wasn't showing up before for some reason. I added a patch at the beginning of the series to break the cycles.
